src/SixLabors.Fonts/TextLayout2.cs

```diff
@@ -0,0 +1,483 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using SixLabors.Fonts.Tables.AdvancedTypographic.Shapers;
+using SixLabors.Fonts.Unicode;
+
+namespace SixLabors.Fonts
+{
+    /// <summary>
+    /// Encapsulated logic or laying out text.
+    /// </summary>
+    internal class TextLayout2
+    {
+        internal static TextLayout2 Default { get; set; } = new TextLayout2();
+
+        public IReadOnlyList<GlyphLayout> GenerateLayout(ReadOnlySpan<char> text, RendererOptions options)
+        {
+            if (text.IsEmpty)
+            {
+                return Array.Empty<GlyphLayout>();
+            }
+
+            if (options.WrappingWidth > 0)
+            {
+                // Trim trailing white spaces from the text
+                text = text.TrimEnd(null);
+            }
+
+            IEnumerable<TextLine> textLines = ProcessText(text, options);
```

Comment:
@tocsoft @brianpopow This is the beginning of a major rewrite of how we layout text separating shaping, breaking from the actual layout code.

The `IEnumerable<TextLine>` contains a collection `glyph-metrics|bidi-run|codepoint-offset` values which should give us enough information to layout bidirectional horizontal text (_vertical if we want also_).

I still have a few things to figure out but I plan to split layout explicitly for LTR and RTL lines choosing the initial direction based upon the directional value of the first bidi run. There are tricky things to consider though.

https://github.com/SixLabors/Fonts/pull/198#issuecomment-920087412 for example is a real head scratcher. Arabic numbers are actually LTR so there's 3 bidi runs in play.